numba/cuda/dispatcher.py

```diff
@@ -688,27 +686,28 @@ def compile_device(self, args):
         Returns the `CompileResult`.
         """
         if args not in self.overloads:
-
-            debug = self.targetoptions.get('debug')
-            inline = self.targetoptions.get('inline')
-            fastmath = self.targetoptions.get('fastmath')
-
-            nvvm_options = {
-                'debug': debug,
-                'opt': 3 if self.targetoptions.get('opt') else 0,
-                'fastmath': fastmath
-            }
-
-            cres = compile_cuda(self.py_func, None, args,
-                                debug=debug,
-                                inline=inline,
-                                fastmath=fastmath,
-                                nvvm_options=nvvm_options)
-            self.overloads[args] = cres
-
-            cres.target_context.insert_user_function(cres.entry_point,
-                                                     cres.fndesc,
-                                                     [cres.library])
+            with self._compiling_counter:
```

Comment:
It's not clear to me that the logic isn't backwards in the base class - shouldn't the base class not be incrementing the compiling counter if it's only going to return an existing overload?

(Or to put it another way - isn't the base class the one with backwards logic, and it too should check for an existing overload prior to incrementing the compiling counter?)

Reply:
@sklam Think you wrote this originally, do you recall why it's this way around? If not, I'll go take a look at what's going on in the base/CPU target, it could be an oversight/bug.

Reply:
To add a little more detail - I think it doesn't technically matter where the check for / return of the existing overload is, because the compiling counter is only used to detect recursion during another compilation that could get started during the first one. Since we won't start compiling if an overload does exist, it doesn't matter whether the compiling counter has been incremented or not (and if it was incremented, it will be decremented again as soon as the existing overload is returned).

Note, I do prefer my variant of the logic where the counter isn't incremented at all if there's an existing overload, because it seems to conflate fewer things at a time.

Reply:
issue at https://github.com/numba/numba/issues/8162